pkg/icingaredis/utils_test.go

```diff
@@ -147,3 +149,201 @@ func TestCreateEntities(t *testing.T) {
+	return &v1.EntityWithChecksum{
+		EntityWithoutChecksum: v1.EntityWithoutChecksum{IdMeta: v1.IdMeta{Id: id}},
+		ChecksumMeta:          v1.ChecksumMeta{PropertiesChecksum: checksum},
+	}
+}
+
+func TestSetChecksums(t *testing.T) {
+	subtests := []struct {
+		name      string
+		input     []database.Entity
+		checksums map[string]database.Entity
+		output    []database.Entity
+		error     bool
+	}{
+		{name: "nil"},
+		{
+			name:      "empty",
+			checksums: map[string]database.Entity{},
+		},
+		{
+			name:      "one",
+			input:     []database.Entity{newTestEntityWithChecksum([]byte{1}, nil, []byte{3})},
+			checksums: map[string]database.Entity{"01": newEntityWithChecksum([]byte{1}, []byte{2})},
+			output:    []database.Entity{newTestEntityWithChecksum([]byte{1}, []byte{2}, []byte{3})},
+		},
+		{
+			name: "two",
+			input: []database.Entity{
+				newTestEntityWithChecksum([]byte{4}, nil, []byte{6}),
+				newTestEntityWithChecksum([]byte{7}, nil, []byte{9}),
+			},
+			checksums: map[string]database.Entity{
+				"04": newEntityWithChecksum([]byte{4}, []byte{5}),
+				"07": newEntityWithChecksum([]byte{7}, []byte{8}),
+			},
+			output: []database.Entity{
+				newTestEntityWithChecksum([]byte{4}, []byte{5}, []byte{6}),
+				newTestEntityWithChecksum([]byte{7}, []byte{8}, []byte{9}),
+			},
+		},
+		{
+			name: "three",
+			input: []database.Entity{
+				newTestEntityWithChecksum([]byte{10}, nil, []byte{12}),
+				newTestEntityWithChecksum([]byte{13}, nil, []byte{15}),
+				newTestEntityWithChecksum([]byte{16}, nil, []byte{18}),
+			},
+			checksums: map[string]database.Entity{
+				"0a": newEntityWithChecksum([]byte{10}, []byte{11}),
+				"0d": newEntityWithChecksum([]byte{13}, []byte{14}),
+				"10": newEntityWithChecksum([]byte{16}, []byte{17}),
+			},
+			output: []database.Entity{
+				newTestEntityWithChecksum([]byte{10}, []byte{11}, []byte{12}),
+				newTestEntityWithChecksum([]byte{13}, []byte{14}, []byte{15}),
+				newTestEntityWithChecksum([]byte{16}, []byte{17}, []byte{18}),
+			},
+		},
+		{
+			name:      "superfluous-checksum",
+			checksums: map[string]database.Entity{"13": newEntityWithChecksum([]byte{19}, []byte{20})},
+		},
+		{
+			name:  "missing-checksum",
+			input: []database.Entity{newTestEntityWithChecksum([]byte{22}, nil, []byte{24})},
+			error: true,
+		},
+	}
+
+	for _, st := range subtests {
+		t.Run(st.name, func(t *testing.T) {
+			for _, concurrency := range []int{1, 2, 30} {
+				t.Run(fmt.Sprint(concurrency), func(t *testing.T) {
+					for _, l := range latencies {
+						t.Run(l.name, func(t *testing.T) {
+							ctx, cancel := context.WithCancel(context.Background())
+							defer cancel()
+
+							input := make(chan database.Entity, 1)
+							go func() {
+								defer close(input)
+
+								for _, v := range st.input {
+									if l.latency > 0 {
+										select {
+										case <-time.After(l.latency):
+										case <-ctx.Done():
+											return
```

Comment:
Indentation depth eleven. We're hitting indentation levels that shouldn't even be possible :>